### PR TITLE
Add full support for OBO user requests

### LINKED ISSUE #509: All requests should use user identity auth if user identity configured
When user configures OBO authentication then some requests authenticated using the user identity of OBO and some continue to use service identity directly.

Properly using user identity:

* **QueryData** ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/e25f75a831651e5825bef183496d778c87624f50/pkg/azuredx/datasource.go#L82))

Methods using service identity directly (need to be fixed):

* CheckHealth([here](https://github.com/grafana/azure-data-explorer-datasource/blob/e25f75a831651e5825bef183496d778c87624f50/pkg/azuredx/datasource.go#L96))
* CallResource **getDatabases** ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/fb3895246c4f2c4e8d2001edaa44e57a06fca8d6/pkg/azuredx/resource_handler.go#L57))
* CallResource **getSchema** ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/fb3895246c4f2c4e8d2001edaa44e57a06fca8d6/pkg/azuredx/resource_handler.go#L28))

Regardless of type of the request, authentication method should be consistent with user's configuration.

### PR DESCRIPTION
**Pull request by aangelisc**
This PR ensures that all requests make use of the user identity if the datasource is configured to do so.

`grafana-plugin-sdk-go` recently added the appropriate support for header forwarding allowing the `X-ID-TOKEN` header to be accessed in external plugins.

Fixes #509 
Fixes grafana/grafana-plugin-sdk-go#579

### CODE REVIEW

**Review by kostrse — APPROVED** [review]
Awesome that a recent version of Plugin SDK already supports headers for CheckHealth.

This fixes #509, but could we keep the https://github.com/grafana/grafana-plugin-sdk-go/issues/579 open since that is a general feature request that would be good to consider by itself?

**Review by andresmgot — APPROVED** [review]
Good job here :+1: 

One slighly weird thing is that if the user creating the data source is not logged with OAuth, the check fails with:

Screenshot from 2023-01-10 11-49-51

I am not sure what would be the best UX here for that situation, we could probably catch the error and show something like "Unable to check data source, the current user has not signed in with OAuth". WDYT?

**Comment by kostrse** [comment]
> One slighly weird thing is that if the user creating the data source is not logged with OAuth, the check fails with:
> 
> Screenshot from 2023-01-10 11-49-51
> 
> I am not sure what would be the best UX here for that situation, we could probably catch the error and show something like "Unable to check data source, the current user has not signed in with OAuth". WDYT?

This error doesn't imply that user didn't use OAuth, it can be other cause like unset auth  toke passthru.

Ideally user should not see this error at all.

I would add a task to intestigate how to show this type of authentication only if OAuth used.

Since it's experimental feature under a feature flag, it's OK if it fails when enabled for non OAuth. It's not enabled by defailt and users which enable it expected to understand scenario when it works.

File: pkg/azuredx/client/client.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func (c *Client) TestRequest(ctx context.Context, datasourceSettings *models.Dat
 	}
 
 	// TODO: This is a workaround because Plugin SDK doesn't expose user context for CheckHealth
```
[the remark this answers is not in the document]
Comment (aangelisc) [reply pkg/azuredx/client/client.go:45]: Thanks!

**Comment by aangelisc** [comment]
> I am not sure what would be the best UX here for that situation, we could probably catch the error and show something like "Unable to check data source, the current user has not signed in with OAuth". WDYT?

My preference would be to actually prevent any requests made if a user signed in with the standard method (service account) attempts to make requests via an OAuth datasource. I think we should maybe implement a catch-all for this situation and return an error along the lines of `Invalid credentials, attempting to use Service Account credentials with OBO authentication enabled`.

**Comment by andresmgot** [comment]
> I would add a task to intestigate how to show this type of authentication only if OAuth was used or to specifically check whether user was signed in using OAuth and show error message which explicitly tells that auth isn't possible becaue OAuth wasn't used.
> 
> Let's keep it to a separate task, this PR fixes check health.

Sounds good to me

> > I am not sure what would be the best UX here for that situation, we could probably catch the error and show something like "Unable to check data source, the current user has not signed in with OAuth". WDYT?
> 
> My preference would be to actually prevent any requests made if a user signed in with the standard method (service account) attempts to make requests via an OAuth datasource. I think we should maybe implement a catch-all for this situation and return an error along the lines of `Invalid credentials, attempting to use Service Account credentials with OBO authentication enabled`.

Also agree, just remember to log the original error so we can debug in case we are masking an unexpected error

**Pull request merged**